zcash_client_sqlite/src/chain.rs

```diff
@@ -64,6 +67,121 @@ where
     Ok(())
 }
 
+/// Data structure representing a row in the block metadata database.
+pub struct BlockMeta {
+    pub height: BlockHeight,
+    pub block_hash: BlockHash,
+    pub block_time: u32,
+    pub sapling_outputs_count: u32,
+    pub orchard_actions_count: u32,
+}
+
+/// Inserts a batch of rows into the block metadata database.
+pub fn blockmetadb_insert(
+    conn: &Connection,
+    block_meta: &[BlockMeta],
+) -> Result<(), SqliteClientError> {
+    let mut stmt_insert = conn.prepare(
+        "INSERT INTO compactblocks_meta (height, blockhash, time, sapling_outputs_count, orchard_actions_count)
+        VALUES (?, ?, ?, ?, ?)"
+    )?;
+
+    conn.execute("BEGIN IMMEDIATE", NO_PARAMS)?;
+    let result = block_meta
+        .iter()
+        .map(|m| {
+            stmt_insert.execute(params![
+                u32::from(m.height),
+                &m.block_hash.0[..],
+                m.block_time,
+                m.sapling_outputs_count,
+                m.orchard_actions_count,
+            ])
+        })
+        .collect::<Result<Vec<_>, _>>();
+    match result {
+        Ok(_) => {
+            conn.execute("COMMIT", NO_PARAMS)?;
+            Ok(())
+        }
+        Err(error) => {
+            match conn.execute("ROLLBACK", NO_PARAMS) {
+                Ok(_) => Err(SqliteClientError::DbError(error)),
+                Err(e) =>
+                    // Panicking here is probably the right thing to do, because it
+                    // means the database is corrupt.
+                    panic!(
+                        "Rollback failed with error {} while attempting to recover from error {}; database is likely corrupt.",
+                        e,
+                        error
+                    )
+            }
+        }
+    }
+}
+
+/// Implements a traversal of `limit` blocks of the filesystem-backed
+/// block cache.
+///
+/// Starting at `from_height`, the `with_row` callback is invoked
+/// with each block retrieved from the backing store. If the `limit`
+/// value provided is `None`, all blocks are traversed up to the
+/// maximum height for which metadata is available.
+pub fn fsblockdb_with_blocks<F, P: AsRef<Path>>(
+    cache: &FsBlockDb<P>,
+    from_height: BlockHeight,
+    limit: Option<u32>,
+    mut with_block: F,
+) -> Result<(), SqliteClientError>
+where
+    F: FnMut(CompactBlock) -> Result<(), SqliteClientError>,
+{
+    // Fetch the CompactBlocks we need to scan
+    let mut stmt_blocks = cache.conn.prepare(
+        "SELECT height, blockhash, time, sapling_outputs_count, orchard_actions_count
+         FROM compactblocks_meta
+         WHERE height > ?
```

Comment:
The doc comment says "starting at `from_height`", which implies `from_height` should be included.
```suggestion
         WHERE height >= ?
```

Reply:
Hm. This is currently written to be consistent with the existing behavior: https://github.com/zcash/librustzcash/blob/main/zcash_client_sqlite/src/chain.rs#L36

I agree that the name of the parameter is misleading, but we should be consistent, and I think it's hazardous to change this now.

Reply:
Renamed to `last_scanned_height` and updated the documentation accordingly.

Reply:
To be clear, there is no error here. The SDK ingress points correctly ensure that this ends up being set to the height of the last validated block, _from which_ we want to scan. English is hard.